drivers/gpio/gpio_airoc.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (c) 2025 Beechwoods Software, Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT infineon_airoc_gpio
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(gpio_airoc, CONFIG_GPIO_LOG_LEVEL);
+
+#include <airoc_whd_hal_common.h>
+whd_interface_t airoc_wifi_get_whd_interface(void);
+whd_result_t whd_wifi_set_iovar_buffer(whd_interface_t ifp, const char *iovar,
+				       void *buffer, uint16_t buffer_length);
+whd_result_t whd_wifi_get_iovar_buffer(whd_interface_t ifp, const char *iovar_name,
+				       uint8_t *out_buffer, uint16_t out_length);
+static whd_interface_t airoc_if;
+static uint32_t gpio_value_buffer;
+
+#define IOVAR_STR_GPIOOUT "gpioout"
+
+#define AIROC_GPIO_PINS 5
+#define AIROC_GPIO_LOGIC1 0x1
+#define AIROC_GPIO_LOGIC0 0x0
+
+#define PSEUDO_IOVAR_GET
+
+struct airoc_gpio_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+};
+
+struct airoc_gpio_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+};
+
+static int airoc_wifi_init(void)
+{
+	int err = 0;
+
+	if (airoc_if != NULL) {
+		return 0;
+	}
+
+	LOG_INF("AIROC initializing...");
+
+	airoc_if = airoc_wifi_get_whd_interface();
+	if (airoc_if == NULL) {
+		LOG_ERR("airoc_if is NULL");
+		err = -ENODEV;
+	}
+
+	LOG_INF("airoc_if: %p", airoc_if);
+
+	return err;
+}
+
+static int airoc_gpio_get(uint32_t *data)
+{
+	uint32_t result;
+
+#ifdef PSEUDO_IOVAR_GET
+	*data = gpio_value_buffer;
+	result = 0;
+#else
+	result = whd_wifi_get_iovar_buffer(airoc_if, IOVAR_STR_GPIOOUT,
+					(uint8_t *)data, sizeof(uint32_t));
+	/* Read back data is always 0U with this undocumented iovar API */
+#endif
+	LOG_INF("%s: data: %d size: %zu", __func__, *data, sizeof(*data));
+	LOG_INF("result: %d", result);
+
+	if (result != WHD_SUCCESS) {
+		LOG_ERR("%s: result: %d", __func__, result);
+		return -EIO;
+	}
+	return 0;
+}
+
+static int airoc_gpio_set(uint8_t gpio_pin, uint8_t gpio_state)
+{
+	uint32_t parameters[] = {gpio_pin, gpio_state};
+	uint32_t result;
+
+	LOG_INF("%s: gpio_pin: %d gpio_state: %d", __func__, gpio_pin, gpio_state);
+
+	result = whd_wifi_set_iovar_buffer(airoc_if, IOVAR_STR_GPIOOUT,
```

Comment:
This design will fundamentally cause a conflict with the Wi-Fi driver,
which will require mutual exclusion.
For devices like this, you'll need to create a placeholder device as an MFD, create a semaphore corresponding to the device instance, and perform mutual exclusion control between each driver (in this case, Wi-Fi and GPIO).

This may be helpful:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/mfd/mfd_maxq10xx.c